File: android/src/com/android/tools/idea/welcome/wizard/deprecated/JdkSetupStep.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.welcome.wizard.deprecated;

import static com.android.tools.idea.gradle.project.AndroidGradleProjectSettingsControlBuilder.ANDROID_STUDIO_DEFAULT_JDK_NAME;
import static com.android.tools.idea.gradle.ui.SdkUiStrings.JDK_LOCATION_TOOLTIP;
import static com.android.tools.idea.gradle.ui.SdkUiStrings.JDK_LOCATION_WARNING_URL;
import static com.android.tools.idea.gradle.ui.SdkUiStrings.generateChooseValidJdkDirectoryError;
import static com.android.tools.idea.gradle.ui.SdkUiUtils.getLocationFromComboBoxWithBrowseButton;
import static com.android.tools.idea.io.FilePaths.toSystemDependentPath;
import static com.android.tools.idea.sdk.IdeSdks.getJdkFromJavaHome;
import static com.android.tools.idea.wizard.WizardConstants.KEY_JDK_LOCATION;
import static com.intellij.openapi.util.io.FileUtilRt.toSystemDependentName;
import static com.intellij.openapi.vfs.VfsUtilCore.virtualToIoFile;

import com.android.tools.idea.flags.StudioFlags;
import com.android.tools.idea.gradle.ui.LabelAndFileForLocation;
import com.android.tools.idea.sdk.IdeSdks;
import com.intellij.ide.BrowserUtil;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileChooser.FileChooserDescriptor;
import com.intellij.openapi.util.SystemInfo;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.ComboboxWithBrowseButton;
import com.intellij.ui.ContextHelpLabel;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.util.Function;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.io.File;
import javax.swing.JComboBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Wizard step for JDK setup.
 * @deprecated use {@link com.android.tools.idea.welcome.wizard.JdkSetupStep}
 */
public class JdkSetupStep extends FirstRunWizardStep {
  @SuppressWarnings("unused") private ComboboxWithBrowseButton myJdkLocationComboBox;
  private JBLabel myJdkLocationHelp;
  private JBScrollPane myContents;
  private boolean myIsJavaHomeValid;

  public JdkSetupStep() {
    super("Select default JDK Location");
    setUpJdkLocationComboBox();
    setComponent(myContents);
    createUIComponents();
  }

  private void setUpJdkLocationComboBox() {
    FileChooserDescriptor descriptor = createSingleFolderDescriptor(file -> {
      File validatedFile = validateJdkPath(file);
      if (validatedFile == null) {
        throw new IllegalArgumentException(generateChooseValidJdkDirectoryError());
      }
      setJdkLocationComboBox(file);
      return null;
    });

    myJdkLocationComboBox.addBrowseFolderListener(getProject(), descriptor);
    JComboBox comboBox = myJdkLocationComboBox.getComboBox();
    IdeSdks ideSdks = IdeSdks.getInstance();
    File embeddedPath = ideSdks.getEmbeddedJdkPath();
    if (embeddedPath != null) {
      File validatedPath = validateJdkPath(embeddedPath);
      if (validatedPath != null) {
        comboBox.addItem(new LabelAndFileForLocation("Embedded JDK", validatedPath));
      }
    }
    String javaHomePath = getJdkFromJavaHome();
    if (javaHomePath != null) {
      File validatedPath = validateJdkPath(new File(javaHomePath));
      myIsJavaHomeValid = validatedPath != null;
      if (myIsJavaHomeValid) {
        comboBox.addItem(new LabelAndFileForLocation("JAVA_HOME", validatedPath));
      }
    }
    comboBox.setEditable(true);
    comboBox.addItemListener(new ItemListener() {
      @Override
      public void itemStateChanged(ItemEvent event) {
        if (event.getStateChange() == ItemEvent.SELECTED) {
          Object selectedItem = event.getItem();
          if (selectedItem instanceof LabelAndFileForLocation) {
            ApplicationManager.getApplication().invokeLater(() -> setJdkLocationComboBox(((LabelAndFileForLocation)selectedItem).getFile()));
          }
        }
      }
    });
    setJdkLocationComboBox(embeddedPath);
  }

  private void setJdkLocationComboBox(@Nullable File path) {
    if (path == null) {
      myJdkLocationComboBox.getComboBox().setSelectedItem(null);
    }
    else {
      myJdkLocationComboBox.getComboBox().setSelectedItem(toSystemDependentName(path.getPath()));
    }
    updateIsValidPath();
  }

  private void createUIComponents() {
    myJdkLocationHelp = ContextHelpLabel.createWithLink(null, JDK_LOCATION_TOOLTIP, "Learn more",
                                                        () -> BrowserUtil.browse(JDK_LOCATION_WARNING_URL));
  }

  @NotNull
  private static FileChooserDescriptor createSingleFolderDescriptor(@NotNull Function<? super File, Void> validation) {
    FileChooserDescriptor descriptor = new FileChooserDescriptor(false, true, false, false, false, false) {
      @Override
      public void validateSelectedFiles(VirtualFile[] files) {
        for (VirtualFile virtualFile : files) {
          File file = virtualToIoFile(virtualFile);
          validation.fun(file);
        }
      }
    };
    if (SystemInfo.isMac) {
      descriptor.withShowHiddenFiles(true);
    }
    descriptor.setTitle("Choose JDK Location");
    return descriptor;
  }

  private void updateIsValidPath() {
    invokeUpdate(null);
  }

  @Nullable
  private File validateJdkPath(@NotNull File file) {
    File possiblePath = IdeSdks.getInstance().validateJdkPath(file);
    if (possiblePath != null) {
      setJdkLocationComboBox(possiblePath);
      return possiblePath;
    }
    return null;
  }

  @Override
  public void init() {
    // Apply default selection
    IdeSdks ideSdks = IdeSdks.getInstance();
    File embeddedPath = ideSdks.getEmbeddedJdkPath();
    setJdkLocationComboBox(embeddedPath);
  }

  @Nullable
  @Override
  public JLabel getMessageLabel() {
    return null;
  }

  @Override
  public JComponent getPreferredFocusedComponent() {
    return myJdkLocationComboBox;
  }

  @Override
  public boolean validate() {
    if (!isValidJdkPath()) {
      return false;
    }
    return super.validate();
  }

  private boolean isValidJdkPath() {
    return validateJdkPath(getJdkLocation()) != null;
  }

  @Override
  public boolean commitStep() {
    if (!isValidJdkPath()) {
      return false;
    }
    File path = toSystemDependentPath(getJdkLocation().getPath());
    if (StudioFlags.ALLOW_JDK_PER_PROJECT.get()) {
      IdeSdks.findOrCreateJdk(ANDROID_STUDIO_DEFAULT_JDK_NAME, path);
    }
    else {
      ApplicationManager.getApplication().runWriteAction(() -> {IdeSdks.getInstance().setJdkPath(path);});
    }
    myState.put(KEY_JDK_LOCATION, path.getPath());
    return true;
  }

  @NotNull
  private File getJdkLocation() {
    return getLocationFromComboBoxWithBrowseButton(myJdkLocationComboBox);
  }

  @Override
  public boolean isStepVisible() {
    return StudioFlags.NPW_SHOW_JDK_STEP.get() && Boolean.TRUE.equals(myState.get(FirstRunWizard.KEY_CUSTOM_INSTALL));
  }
}

File: android/src/com/android/tools/idea/welcome/wizard/JdkSetupStep.kt
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.welcome.wizard

import com.android.tools.adtui.validation.Validator
import com.android.tools.adtui.validation.ValidatorPanel
import com.android.tools.idea.gradle.ui.LabelAndFileForLocation
import com.android.tools.idea.gradle.ui.SdkUiStrings.generateChooseValidJdkDirectoryError
import com.android.tools.idea.gradle.ui.SdkUiUtils.getLocationFromComboBoxWithBrowseButton
import com.android.tools.idea.observable.core.BoolValueProperty
import com.android.tools.idea.observable.core.ObservableBool
import com.android.tools.idea.observable.core.StringValueProperty
import com.android.tools.idea.sdk.IdeSdks
import com.android.tools.idea.sdk.IdeSdks.ANDROID_STUDIO_DEFAULT_JDK_NAME
import com.android.tools.idea.ui.validation.validators.PathValidator
import com.android.tools.idea.wizard.model.ModelWizardStep
import com.android.tools.idea.wizard.ui.WizardUtils.wrapWithVScroll
import com.intellij.openapi.fileChooser.FileChooserDescriptor
import com.intellij.openapi.util.SystemInfo
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.ui.ComboboxWithBrowseButton
import com.intellij.ui.layout.panel
import com.intellij.uiDesigner.core.Spacer
import java.awt.event.ItemEvent
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Wizard step for JDK setup.
 */
class JdkSetupStep(model: FirstRunModel) : ModelWizardStep<FirstRunModel>(model, "Select default JDK location") {
  private val jdkLocationComboBox = ComboboxWithBrowseButton()

  private val jdkPanel = panel {
    row {
      label("Select the directory where the Java Development Kit (JDK) is located.")
    }
    row {
      Spacer()()
    }
    row {
      jdkLocationComboBox()
    }
  }

  private val validatorPanel = ValidatorPanel(this, wrapWithVScroll(jdkPanel))
  private val invalidPathMessage = StringValueProperty()
  private val isValidJdkPath = BoolValueProperty(false)
  private val jdkLocation: Path get() = getLocationFromComboBoxWithBrowseButton(jdkLocationComboBox)

  init {
    val descriptor = createSingleFolderDescriptor { file ->
      validateJdkPath(file) ?: throw IllegalArgumentException(generateChooseValidJdkDirectoryError())
      setJdkLocationComboBox(file)
    }

    jdkLocationComboBox.addBrowseFolderListener(null, descriptor)

    val comboBox = jdkLocationComboBox.comboBox
    comboBox.addActionListener {
      validateJdkPath(jdkLocation)
    }

    fun addJdkIfValid(path: Path?, label: String) {
      path ?: return
      val validatedPath = validateJdkPath(path) ?: return
      comboBox.addItem(LabelAndFileForLocation(label, validatedPath))
    }

    val embeddedPath = IdeSdks.getInstance().embeddedJdkPath
    addJdkIfValid(embeddedPath, "Embedded JDK")

    val javaHomePath = IdeSdks.getJdkFromJavaHome()
    if (javaHomePath != null) {
      addJdkIfValid(Paths.get(javaHomePath), "JAVA_HOME")
    }

    comboBox.isEditable = true
    comboBox.addItemListener { event ->
      val selectedItem = event.item
      if (event.stateChange == ItemEvent.SELECTED && selectedItem is LabelAndFileForLocation) {
        invokeLater { setJdkLocationComboBox(selectedItem.file) }
      }
    }
    setJdkLocationComboBox(embeddedPath)
  }

  override fun onEntering() {
    validatorPanel.registerMessageSource(invalidPathMessage)
  }

  override fun canGoForward(): ObservableBool = isValidJdkPath

  override fun onProceeding() {
    val path = jdkLocation.toAbsolutePath().normalize()
    IdeSdks.findOrCreateJdk(ANDROID_STUDIO_DEFAULT_JDK_NAME, path!!)
  }

  override fun getPreferredFocusComponent() = jdkPanel

  override fun getComponent() = validatorPanel

  private fun validateJdkPath(file: Path): Path? {
    val possiblePath = IdeSdks.getInstance().validateJdkPath(file)
    if (possiblePath != null) {
      setJdkLocationComboBox(possiblePath)
      isValidJdkPath.set(true)
      return possiblePath
    }
    isValidJdkPath.set(false)
    return null

    // TODO(qumeric): replace it with PathValidator, like:
    val validator = PathValidator.Builder().withCommonRules().build("Android SDK location")
    val validationResult = validator.validate(jdkLocation)
    invalidPathMessage.set(validationResult.message)
    val isError = validationResult.severity != Validator.Severity.ERROR
    isValidJdkPath.set(isError)
    return file.takeIf { !isError }
  }

  private fun setJdkLocationComboBox(path: Path?) {
    jdkLocationComboBox.comboBox.selectedItem = path?.toString()
  }
}

// TODO(qumeric) make private
fun createSingleFolderDescriptor(validation: (Path) -> Unit): FileChooserDescriptor {
  val result = object : FileChooserDescriptor(false, true, false, false, false, false) {
    override fun validateSelectedFiles(files: Array<VirtualFile>) {
      files.map(VirtualFile::toNioPath).forEach(validation)
    }
  }
  result.withShowHiddenFiles(SystemInfo.isMac)
  result.title = "Choose JDK Location"
  return result
}
